src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -309,8 +309,12 @@ QgsSymbolLegendNode::QgsSymbolLegendNode( QgsLayerTreeLayer *nodeLayer, const Qg
   }
 
   updateLabel();
+
   if ( QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( nodeLayer->layer() ) )
+  {
     connect( vl, &QgsVectorLayer::symbolFeatureCountMapChanged, this, &QgsSymbolLegendNode::updateLabel );
+    connect( vl, &QgsVectorLayer::rendererChanged, nodeLayer, &QgsLayerTreeLayer::updateSymbolExpressions );
```

Comment:
I have no clue why but this fires CONSTANTLY with rule-based symbology. This explain a good part of the lag. I'll see if I can find a better way to track symbology changes.